physical/raft/snapshot.go

```diff
@@ -151,69 +155,153 @@ func (f *BoltSnapshotStore) getBoltSnapshotMeta() (*raft.SnapshotMeta, error) {
 	}
 
 	if latestConfig != nil {
-		index, configuration := protoConfigurationToRaftConfiguration(latestConfig)
-		meta.Configuration = configuration
-		meta.ConfigurationIndex = index
+		meta.ConfigurationIndex, meta.Configuration = protoConfigurationToRaftConfiguration(latestConfig)
 	}
 
 	return meta, nil
 }
 
 // Open takes a snapshot ID and returns a ReadCloser for that snapshot.
 func (f *BoltSnapshotStore) Open(id string) (*raft.SnapshotMeta, io.ReadCloser, error) {
-	var readCloser io.ReadCloser
-	var meta *raft.SnapshotMeta
-	switch id {
-	case boltSnapshotID:
-
-		var err error
-		meta, err = f.getBoltSnapshotMeta()
-		if err != nil {
-			return nil, nil, err
-		}
-		// If we don't have any data return an error
-		if meta.Index == 0 {
-			return nil, nil, errors.New("no snapshot data")
-		}
+	if id == boltSnapshotID {
+		return f.openFromFSM()
+	}
 
-		// Stream data out of the FSM to calculate the size
-		var writeCloser *io.PipeWriter
-		readCloser, writeCloser = io.Pipe()
-		metaReadCloser, metaWriteCloser := io.Pipe()
-		go func() {
-			f.fsm.writeTo(context.Background(), metaWriteCloser, writeCloser)
-		}()
-
-		// Compute the size
-		n, err := io.Copy(ioutil.Discard, metaReadCloser)
-		if err != nil {
-			f.logger.Error("failed to read state file", "error", err)
-			metaReadCloser.Close()
-			readCloser.Close()
-			return nil, nil, err
-		}
+	return f.openFromFile(id)
+}
 
-		meta.Size = n
+func (f *BoltSnapshotStore) openFromFSM() (*raft.SnapshotMeta, io.ReadCloser, error) {
+	meta, err := f.getMetaFromFSM()
+	if err != nil {
+		return nil, nil, err
+	}
+	// If we don't have any data return an error
+	if meta.Index == 0 {
+		return nil, nil, errors.New("no snapshot data")
+	}
 
-	default:
-		var err error
-		meta, readCloser, err = f.fileSnapStore.Open(id)
-		if err != nil {
-			return nil, nil, err
+	// Stream data out of the FSM to calculate the size
+	readCloser, writeCloser := io.Pipe()
+	metaReadCloser, metaWriteCloser := io.Pipe()
+	go func() {
+		f.fsm.writeTo(context.Background(), metaWriteCloser, writeCloser)
+	}()
+
+	// Compute the size
+	n, err := io.Copy(ioutil.Discard, metaReadCloser)
+	if err != nil {
+		f.logger.Error("failed to read state file", "error", err)
+		metaReadCloser.Close()
+		readCloser.Close()
+		return nil, nil, err
+	}
+
+	meta.Size = n
+
```

Comment:
Do we need to close `metaReadCloser` before returning? I'm not sure how `metaReadCloser` and `readCloser` are linked.